### ISSUE TITLE
arrow-parens: as-needed doesn't account for destructuring arguments

### ISSUE THREAD

**Issue by hawkrives** [eslint/eslint#3353]
```js
let x = ({item}) => item
let y = ([item]) => item
```

Both of these require parentheses, but `arrow-parens: as-needed` reports them as not needing parens.

BabelJS repl tests: [object destructuring], and [array destructuring].

[object destructuring]: http://babeljs.io/repl/#?experimental=true&evaluate=false&loose=false&spec=false&code=let%20x%20%3D%20%7Bitem%7D%20%3D%3E%20item%0A

[array destructuring]: http://babeljs.io/repl/#?experimental=true&evaluate=false&loose=false&spec=false&code=let%20y%20%3D%20%5Bitem%5D%20%3D%3E%20item%0A

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by hawkrives** [comment]
Oh yes, this comes about as a result of #3277. I like the rule, it's just a bit ... overzealous.

**Comment by xjamundx** [comment]
Any ideas of other cases that might be missing?

**Comment by xjamundx** [comment]
There are also a number of typos in the docs which I'll sneak into this fix :)

**Comment by hawkrives** [comment]
@xjamundx: I don't think there are any others—unless there's a third type of restructuring syntax. Thanks for working on this!

**Comment by hawkrives** [comment]
Thanks!
